lsp/nls/src/requests/completion.rs

```diff
@@ -21,42 +29,194 @@ pub fn handle_completion(
+                server.reply(Response::new_ok(id, Value::Null));
+                return Ok(());
+            }
+            let name_id = name_id.unwrap();
+
+            linearization
+                .get_in_scope(&item)
+                .iter()
+                .filter_map(|i| {
+                    let (ty, _) = linearization.resolve_item_type_meta(i);
+                    match i.kind {
+                        // Get record fields from static type info
+                        TermKind::Declaration(..)
+                            if name_id == i.id && matches!(ty, Types(AbsType::Record(..))) =>
+                        {
+                            match &ty {
+                                Types(AbsType::Record(row)) => {
+                                    Some((extract_ident(row), i.ty.clone()))
+                                }
+                                _ => unreachable!(),
+                            }
+                        }
+                        // Get record fields from contract metadata
+                        TermKind::Declaration(_, _, ValueState::Known(body_id))
+                            if name_id == i.id
+                                && find_record_contract(&linearization.linearization, body_id)
+                                    .is_some() =>
+                        {
+                            // unwrapping is safe here because of the pattern condition.
+                            let contract =
+                                find_record_contract(&linearization.linearization, body_id)
+                                    .unwrap();
+                            let fields = match &contract.types {
+                                Types(AbsType::Record(row)) => extract_ident(row),
+                                _ => Vec::with_capacity(0),
+                            };
+                            Some((fields, i.ty.clone()))
+                        }
+
+                        // Get record fields from lexical scoping
+                        TermKind::Declaration(_, _, ValueState::Known(body_id))
+                            if name_id == i.id =>
+                        {
+                            find_record_fields(&linearization.linearization, body_id)
+                                .map(|idents| Some((idents, i.ty.clone())))
+                                .unwrap_or(None)
+                        }
+
+                        _ => None,
+                    }
+                })
+                .collect()
+        }
+
+        // variable name completion
+        Some(..) | None => linearization
+            .get_in_scope(&item)
+            .iter()
+            .filter_map(|i| match i.kind {
+                TermKind::Declaration(ref ident, _, _) => Some((vec![ident.clone()], i.ty.clone())),
+                _ => None,
+            })
+            .collect::<Vec<_>>(),
+    };
+
+    // TODO: remove duplicates
```

Comment:
This TODO should probably be addressed in this PR, or an issue should be created and referenced here to track this remaining work